src/middlewared/middlewared/plugins/kmip.py

```diff
@@ -0,0 +1,384 @@
+                try:
+                    uid = self.__register_secret_data(self.zfs_keys[ds['name']], conn)
+                except Exception:
+                    failed.append(ds['name'])
+                    update_data = {'kmip_uid': None} if destroy_successful else {}
+                else:
+                    update_data = {'encryption_key': None, 'kmip_uid': uid}
+                if update_data:
+                    self.middleware.call_sync('datastore.update', zfs_datastore, ds['id'], update_data)
+        return failed
+
+    @private
+    def sync_zfs_keys_from_server_to_db(self, ids=None):
+        zfs_datastore = self.middleware.call_sync('pool.dataset.dataset_datastore')
+        datasets = self.middleware.call_sync(
+            'datastore.query', self.middleware.call_sync('pool.dataset.dataset_datastore'), [
+                ['kmip_uid', '!=', None], ['id', 'in' if ids else 'nin', ids or []]
+            ]
+        )
+        existing_datasets = {ds['name']: ds for ds in self.middleware.call_sync('pool.dataset.query')}
+        failed = []
+        with self.connection() as conn:
+            for ds in filter(lambda d: d['name'] in existing_datasets, datasets):
+                try:
+                    if ds['name'] in self.zfs_keys and self.middleware.call_sync(
+                        'zfs.dataset.check_key', ds['name'], {'key': self.zfs_keys[ds['name']]}
+                    ):
+                        key = self.zfs_keys[ds['name']]
+                    else:
+                        key = self.__retrieve_secret_data(ds['kmip_uid'], conn)
+                except Exception:
+                    failed.append(ds['name'])
+                else:
+                    update_data = {'encryption_key': self.middleware.call_sync('pwenc.encrypt', key), 'kmip_uid': None}
+                    self.middleware.call_sync('datastore.update', zfs_datastore, ds['id'], update_data)
+                    self.zfs_keys.pop(ds['name'], None)
+                    self.__revoke_and_destroy_key(ds, conn)
+        self.zfs_keys = {k: v for k, v in self.zfs_keys.items() if k in existing_datasets}
+        return failed
+
+    @private
+    @accepts(List('ids', null=True, default=[]))
+    @job(lock=lambda args: f'sync_zfs_keys_{args}')
+    def sync_zfs_keys(self, job, ids=None):
+        config = self.middleware.call_sync('kmip.config')
+        if not self.middleware.call_sync('kmip.zfs_keys_pending_sync') or not self.test_connection_and_alert():
+            return
+        if config['enabled'] and config['manage_zfs_keys']:
+            failed = self.sync_zfs_keys_from_db_to_server(ids)
+        else:
+            failed = self.sync_zfs_keys_from_server_to_db(ids)
+        if failed:
+            self.middleware.call_sync(
+                'alert.oneshot_create', ' KMIPZFSDatasetsSyncFailure', {'datasets': ','.join(failed)}
+            )
+
+    @periodic(interval=86400)
+    @job(lock='sync_kmip_keys')
```

Comment:
Lock is not necessary - sync_zfs_keys protects itself. Also it will be better to name it `periodic_sync_keys` so users can understand what is the difference between `sync_zfs_keys` and `sync_keys`

Reply:
`sync_zfs_keys` i think is better wrt API because it's not only a periodic job and users can initiate this themselves as well to get remote server and local db in sync. Please advise if you think we should still rename it, I'll take care of it then